src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.Tasks/WriteSbrpUsageReport.cs

```diff
@@ -0,0 +1,195 @@
+{
+    private const string SbrpRepoName = "source-build-reference-packages";
+
+    private readonly Dictionary<string, PackageInfo> _sbrpPackages = [];
+
+    /// <summary>
+    /// Path to the VMR src directory.
+    /// </summary>
+    [Required]
+    public string SrcPath { get; set; }
+
+    /// <summary>
+    /// Path to the usage report to.
+    /// </summary>
+    [Required]
+    public string OutputPath { get; set; }
+
+    public override bool Execute()
+    {
+        Log.LogMessage($"Scanning for SBRP Package Usage...");
+
+        ReadSbrpPackages("referencePackages", trackTfms: true);
+        ReadSbrpPackages("textOnlyPackages", trackTfms: false);
+
+        ScanProjectReferences();
+
+        GenerateUsageReport();
+
+        return !Log.HasLoggedErrors;
+    }
+
+    private void GenerateUsageReport()
+    {
+        Report report = new();
+        report.Sbrps = [.. _sbrpPackages.Values.OrderBy(pkg => pkg.Id)];
+        PurgeNonReferencedReferences();
+        report.UnreferencedSbrps = [.. GetUnreferencedSbrps().Select(pkg => pkg.Id).OrderBy(id => id)];
+
+        string reportFilePath = Path.Combine(OutputPath, "sbrpPackageUsage.json");
+#pragma warning disable CA1869 // Cache and reuse 'JsonSerializerOptions' instances
+        string jsonContent = JsonSerializer.Serialize(report, new JsonSerializerOptions { WriteIndented = true });
+#pragma warning restore CA1869 // Cache and reuse 'JsonSerializerOptions' instances
+        File.WriteAllText(reportFilePath, jsonContent);
+    }
+
+    /// <summary>
+    /// Removes all references from unreferenced SBRP packages. This is necessary to determine the
+    /// complete set of unreferenced SBRP packages.
+    /// </summary>
+    private void PurgeNonReferencedReferences()
+    {
+        bool hasPurged;
+        do
+        {
+            hasPurged = false;
+            PackageInfo[] unrefPkgs = GetUnreferencedSbrps()
+                .Select(pkg => pkg)
+                .ToArray();
+
+            foreach (PackageInfo sbrpPkg in _sbrpPackages.Values)
+            {
+                foreach (PackageInfo unrefPkg in unrefPkgs)
+                {
+                    var unref = sbrpPkg.References.Keys
+                        .SingleOrDefault(path => path.Contains(SbrpRepoName) && path.Contains($"{unrefPkg.Name}.{unrefPkg.Version}"));
+                    if (unref != null)
+                    {
+                        Log.LogMessage($"Removing {unrefPkg.Id} from {sbrpPkg.Id}'s references.");
+                        sbrpPkg.References.Remove(unref);
+                        hasPurged = true;
+                    }
+                }
+            }
+        } while (hasPurged);
+    }
+
+    private IEnumerable<PackageInfo> GetUnreferencedSbrps() =>
+        _sbrpPackages.Values.Where(pkg => pkg.References.Count == 0);
+
+    private string GetSbrpPackagesPath(string packageType) =>
+        Path.Combine(SrcPath, SbrpRepoName, "src", packageType, "src");
```

Comment:
Consider avoiding this hardcode. We already pass the sbrp repo path in into other msbuild tasks, i.e. https://github.com/dotnet/sdk/blob/91931824692073c445f634e29f3af8ceab2d04a0/src/SourceBuild/content/repo-projects/Directory.Build.targets#L190